sdk/trace/src/main/java/io/opentelemetry/sdk/trace/TracerSharedState.java

```diff
@@ -79,6 +87,15 @@ SpanProcessor getActiveSpanProcessor() {
     return activeSpanProcessor;
   }
 
+  TracerConfig getTracerConfig(InstrumentationScopeInfo instrumentationScopeInfo) {
+    for (ScopeSelector scopeSelector : tracerConfigMap.keySet()) {
+      if (scopeSelector.matchesScope(instrumentationScopeInfo)) {
+        return tracerConfigMap.get(scopeSelector);
```

Comment:
I'm thinking about the places where we provide "customizers" for tracer/meter/logger provider builders, and I'm wondering how many troubles (if any) could happen if two or more customizers added config selectors that overlap, given that only one config will be picked. It definitely seems to be a tricky issue to solve though, but I still would like to see if there might be ways to reduce or eliminate the unexpected behaviors that this situation could bring, maybe by combining here all the matching configs (similar to what happens when multiple SpanProcessors are added to the same tracerprovider)? - Which is something that could work for the `isEnabled` config option use case though I guess it's difficult to ensure that it would work well for other types of config options that might be added in the future too, so that's why I believe it's a tricky thing to address.

Nevertheless, I don't think this is a blocker, I really like this approach the way it is now and it seems like it allows for dynamic configs as well so I'm looking forward to it. Just wanted to mention that potential issue, which I'm not even sure might actually happen, or happen enough times to become a big problem, just to check if I'm missing something here or if someone can come up with a better scalable alternative to the one I mentioned.

Reply:
No you make a good point and its something I've thought about as well.

Consider someone who wants to have a distribution which disables all scopes by default, and wants to re-enable them on a case by case basis. This seems like a perfectly reasonable thing to do, but isn't possible with the simplistic first matching wins semantics I've prototyped here.

Reply:
Got it, I was just thinking further about it, what if the configs weren't tied to a scope and instead got the `InstrumentationScopeInfo` as a parameter on every config function?

So for example, we'd add configs like so:

```java
SdkTracerProvider.builder()
                .addScopeConfig(TracerConfig.disabled())
                .build())
```

And then the config would look like the following:

```java
public abstract class TracerConfig {
  //...
  public abstract boolean isEnabled(InstrumentationScopeInfo scope);
}
```

That way we could also be able to evaluate each config differently where it's read, depending on the case, e.g. for the `isEnabled` one, we could loop through all the available configs until we found one that returned `false`, making the span not to get created, or to move forward with the span creation in case no config had disabled the provided scope. And, for the case of the distribution use case, a distribution could provide a single config where it can change its decision dynamically to cover as many or as few (or zero) scopes within its `isEnabled` impl.

Reply:
The evaluation of whether a tracer is enabled will happen on the creation of every span, so we want it to be very fast. The idea you suggest requires evaluation of the rules on the hot path. I think we need to be able to evaluate the rules off the hot path, and be able to update each SdkTracer with a property indicating whether its enabled or disabled so that evaluation is fast.

Reply:
Ah, fair enough, I wasn't considering the performance penalty.

Reply:
Could do something like:
```
public interface TracerConfigProvider {
  public TracerConfig getConfig(InstrumentationScopeInfo);

  public static TracerConfigProvider disableAllExcept(List<InstrumentationScopeInfo> enabledScopes) {...}
  public static TracerConfigProvider enableAllExcept(List<InstrumentationScopeInfo> disabledScopes) {...}

}
```
And:
- Evaluate `TracerConfigProvider#getConfig(InstrumenationSopeInfo)` every time a tracer is created
- Provide a method for updating the `TracerConfigProvider` called `SdkTracerProvider#updateTracerConfigProvider`
- When `SdkTracerProvider#updateTracerConfigProvider`, revaluate `TracerConfigProvider#getConfig(InstrumentationScopeInfo)` for every tracer that has been created up till that point.

Reply:
I think the ability to change some tracer/meter/logger-provider-wide config somehow at any time should be the way to go to avoid the performance issues (and probably even the multiple overlapping configs too?) So from a first glance the `SdkTracerProvider#updateTracerConfigProvider` part makes sense to me. I'm no longer at work though so I can't take a thorough look right now.

Reply:
Probably having a setter like:

```java
SdkTracerProvider.builder()
    .setTracerConfigProvider(TracerConfigProvider.getDefault())
    .build();
```
Should that be enough without having to provide a `SdkTracerProvider#updateTracerConfigProvider` too? - I'm mostly thinking about the dynamic config use-case and I think it should be enough to provide my own `TracerConfigProvider` once and change its response dynamically later, so maybe it's not needed to provide a setter for a new `TracerConfigProvider` after the TracerProvider is built?

For when the config provider changes its response and we'd like to update an existing Tracer that's using the old config, maybe we could create a pub/sub mechanism to notify the creator of said Tracer to create a new one with the new config.

Reply:
> Should that be enough without having to provide a SdkTracerProvider#updateTracerConfigProvider too? 

Not sure. Couple potential issues:
- `TracerConfigProvider` would have to be an interface such that its possible to have an implementation which is dynamic. 
- I mentioned above that we would evaluate `TracerConfigProvider#getConfig(InstrumentationScopeInfo)` each time a tracer is created. So changing the implementation of `TracerConfigProvider#getConfig(..)` would only affect the Tracers affected after the update. We need some sort of signal that we need to recompute `TracerConfigProvider#getConfig(..)` for all existing Tracers. Calling a `SdkTracerProvider#updateTracerConfigProvider` is one way to provide that signal.

Reply:
Got it, if I understood correctly, It seems like we're talking about the same type of solution composed by:

- Single config provider.
- Config evaluation on tracer/logger/meter creation.
- Observer pattern to notify existing tracer/logger/meters when the config has changed.

But the implementations are different, the one I'm talking about I think could have `TracerConfigProvider` as an interface as you mentioned, and then an abstract implementation of it, maybe called `ObservableTracerConfigProvider`, where there would be an `ObservableTracerConfigProvider#notifyConfigChange` protected method which could serve as the signal to update the existing Tracers. This would place all the responsibility on the TracerConfigProvider impl to make the dynamic config work properly though, which I'm not sure if it's a good or a bad thing, and I guess it could be annoying having to check the type of the configured TracerConfigProvider to verify its the observable one before subscribing to config changes.

`SdkTracerProvider#updateTracerConfigProvider` works too though, it's just that it seems uncommon to have setters for already built tracer/meter/logger providers and I'm not sure if that "mutability" might cause some unexpected behavior scenarios, i.e. somebody can call `GlobalOpenTelemetry.get().getTracerProvider().updateTracerConfigProvider`, even from a library, and turn all the traces off without the host project being aware of it.

Reply:
> SdkTracerProvider#updateTracerConfigProvider works too though, it's just that it seems uncommon to have setters for already built tracer/meter/logger providers 

I think we've resisted these because up till this point we've punted on dynamic configuration, but once we cross that bridge we'll have no choice but to add setters / update methods on the provider classes.

> somebody can call GlobalOpenTelemetry.get().getTracerProvider().updateTracerConfigProvider, even from a library, and turn all the traces off without the host project being aware of it.

Not possible (or rather difficult). `GlobalOpenTelemetry` obfuscates the underlying `OpenTelemetrySdk` instance such that you can't access it without using reflection. And if someone is willing to use reflection, then not having setters / update methods won't be an effective deterrent either.

Reply:
> I think we've resisted these because up till this point we've punted on dynamic configuration, but once we cross that bridge we'll have no choice but to add setters / update methods on the provider classes.

Gotcha.

> Not possible (or rather difficult). GlobalOpenTelemetry obfuscates the underlying OpenTelemetrySdk instance such that you can't access it without using reflection. And if someone is willing to use reflection, then not having setters / update methods won't be an effective deterrent either.

Ah, I see! Yeah definitely if someone wanted to use reflection it wouldn't matter anyway. Now that you mentioned the obfuscation, I was taking a look at the current OTel impl and doing some dry runs where I needed to provide dynamic config changes, and I noticed that it seems like the `[Singal]Provider`s are also obfuscated inside `OpenTelemetrySdk`, which would mean that even if I had my OTel instance handy (without having to get it from `GlobalOpenTelemetry`), I still wouldn't be able to cast the providers to later call their setters. So I guess the way to make the dynamic config work would require storing the `[Signal]Provider`s somewhere else during the building of the `OpenTelemetrySdk` instance, and I just wanted to confirm if that's the expected behavior? I'm asking just to get as close as possible to the full picture of what we'd have to do in OTel Android to allow users and/or vendors to provide dynamic configs, since the OTel instance creation is done internally [in the Android lib](https://github.com/open-telemetry/opentelemetry-android/blob/4d472015745d9d093f4ef29c09d488456af8194e/instrumentation/src/main/java/io/opentelemetry/android/OpenTelemetryRumBuilder.java#L64), I guess similarly to what's done in `AutoConfiguredOpenTelemetrySdkBuilder`, so, in this case, we'd have to provide the real `[Signal]Provider`s somehow over there after the OTel instance is created. Although we currently [have a builder option](https://github.com/open-telemetry/opentelemetry-android/blob/4d472015745d9d093f4ef29c09d488456af8194e/instrumentation/src/main/java/io/opentelemetry/android/OpenTelemetryRum.java#L62) that receives a previously created `OpenTelemetry` instance, in case some users would like to control its creation and we later take it to send telemetry for them, though in this scenario it seems like the dynamic config would not be possible as is.

Btw I appreciate your patience with all my comments, I know they're a lot, is just that I'm trying to come up with everything that might look like a potential problem for my use case early, though hopefully all of them are false positives or have alternative solutions.